File: frontend/src/api/factorCoverage.ts
import { apiClient, type RequestOptions } from './client'

export interface FactorCoverageData {
  strategy_name: string
  total_factors: number
  available_factors: string[]
  stub_factors: string[]
  coverage_rate: number
  configured_weights: Record<string, number>
  effective_weights: Record<string, number>
  weight_drift: Record<string, number>
}

export interface AllCoverageData {
  strategies: FactorCoverageData[]
  global_stub_factors: string[]
}

/**
 * 获取策略因子覆盖率。
 * Get factor coverage for a strategy.
 */
export async function getFactorCoverage(
  name: string,
  options?: RequestOptions
): Promise<FactorCoverageData> {
  const { data } = await apiClient.get(`/strategies/${name}/factor-coverage`, {
    ...(options?.silent ? { __silent: true } : {}),
  } as any)
  return data
}

/**
 * 获取全部策略因子覆盖率。
 * Get factor coverage for all strategies.
 */
export async function getAllFactorCoverage(options?: RequestOptions): Promise<AllCoverageData> {
  const { data } = await apiClient.get('/strategies/factor-coverage-all', {
    ...(options?.silent ? { __silent: true } : {}),
  } as any)
  return data
}

File: backend/app/api/v1/schemas/factor_coverage.py
"""
因子覆盖率 API 响应模型。
Factor coverage API response models.
"""

from pydantic import BaseModel, Field


class FactorCoverageResponse(BaseModel):
    """单个策略的因子覆盖率 / Factor coverage for one strategy."""

    strategy_name: str = Field(description="策略名称")
    total_factors: int = Field(description="配置的因子总数")
    available_factors: list[str] = Field(description="可用因子列表")
    stub_factors: list[str] = Field(description="Stub (未实现) 因子列表")
    coverage_rate: float = Field(
        description="覆盖率 0.0-1.0",
        examples=[0.5556],
    )
    configured_weights: dict[str, float] = Field(description="YAML 配置权重 (原始值)")
    effective_weights: dict[str, float] = Field(description="实际生效权重 (归一化后)")
    weight_drift: dict[str, float] = Field(description="权重漂移量 = effective - configured")


class AllCoverageResponse(BaseModel):
    """全部策略的因子覆盖率 / Factor coverage for all strategies."""

    strategies: list[FactorCoverageResponse]
    global_stub_factors: list[str] = Field(description="全局 stub 因子清单")
